strinject.c

```diff
@@ -0,0 +1,462 @@
+		}
+
+		if (inject_type == STRINJECT_TYPE_UNKNOWN)
+			continue;
+
+		/* get function list */
+		for (i = 0; i < STRINJECT_FUNC_MAX; i++) {
+			if (rtr_check_config_token(g_inject_funcs[i].name, func_list, "|", &reverse) &&
+				!g_strinject_infos[i].exist) {
+				g_strinject_infos[i].type = inject_type;
+				real_strcpy(g_strinject_infos[i].param, inject_param);
+				g_strinject_infos[i].rate = inject_rate;
+
+				g_strinject_infos[i].exist = 1;
+			}
+		}
+	}
+
+	/* set init flag */
+	g_strinject_init = 1;
+}
+
+/*
+ * parse injection parameter
+ */
+
+static int parse_inject_param(enum RTR_STRINJECT_TYPE type, const char *param, size_t len, void *val, off_t *offset)
+{
+	char param_val[512];
+	char *p;
+
+	int fmt_count, overflow_buf_len;
+
+	/* parse param */
+	p = real_strchr(param, ':');
+	if (!p)
+		return -1;
+
+	memset(param_val, 0, sizeof(param_val));
+	real_strncpy(param_val, param, p - param);
+
+	if (real_strcmp(p + 1, "RANDOM") == 0)
+		*offset = rand() % len;
+	else {
+		*offset = strtol(p + 1, NULL, 10);
+		if (*offset == LONG_MIN || *offset == LONG_MAX)
+			return -1;
+	}
+
+	/* check offset */
+	if (*offset >= len)
+		*offset = len - 1;
+
+	/* inject string by types */
+	switch (type) {
+	case STRINJECT_TYPE_HEX:
+		if (real_strcmp(param_val, "RANDOM") == 0)
+			((char *) val)[0] = rand() % 0xFF;
+		else if (real_strncmp(param_val, "0x", 2) == 0) {
+			char hex_val;
+
+			hex_val = strtol(param_val + 2, NULL, 16);
+			if (hex_val == LONG_MIN || hex_val == LONG_MAX)
+				return -1;
+
+			((char *) val)[0] = hex_val;
+		} else
+			return -1;
+
+		break;
+
+	case STRINJECT_TYPE_FMT_STR:
+		fmt_count = strtol(param_val, NULL, 10);
+		if (fmt_count == LONG_MIN || fmt_count == LONG_MAX)
+			return -1;
+
+		if (fmt_count > (len - *offset) / 2)
+			fmt_count = (len - *offset) / 2;
+
+		*((int *) val) = fmt_count;
+
+		break;
+
+	case STRINJECT_TYPE_BUF_OVERFLOW:
+		overflow_buf_len = strtol(param_val, NULL, 10);
+		if (overflow_buf_len == LONG_MIN || overflow_buf_len == LONG_MAX)
+			return -1;
+
+		if (overflow_buf_len > (len - *offset))
+			overflow_buf_len = (len - *offset);
+
+		*((int *) val) = overflow_buf_len;
+
+		break;
+
+	case STRINJECT_TYPE_FILE_LINE:
+		real_strncpy((char *) val, param_val, real_strlen(param_val));
+		break;
+	}
+
+	return 0;
+}
+
+/*
+ * inject single hex value
+ */
+
+int inject_single_hex(const char *param, void *buffer, size_t len)
+{
+	char hex_value;
+	off_t offset;
+
+	char *p;
+
+	/* parse injection parameter */
+	if (parse_inject_param(STRINJECT_TYPE_HEX, param, len, (void *) &hex_value, &offset) != 0)
+		return 0;
+
+	/* allocate memory for buffer */
+	p = (char *) real_malloc(len);
+	if (!p)
+		return 0;
+
+	/* set destination buffer */
+	if (offset > 0)
+		real_memcpy(p, buffer, offset);
+
+	real_memcpy(&p[offset], &hex_value, 1);
+	if (offset < len)
+		real_memcpy(p + offset + 1, buffer + offset + 1, len - offset - 1);
+
+	real_memcpy(buffer, p, len);
+
+	/* free buffer */
+	real_free(p);
```

Comment:
Could lines 222 - 238 be condensed to `char *p = buffer; p[offset] = hex_value;` or `((char *)buffer)[offset] = hex_value`?